### ISSUE TITLE
Partials ignore context if used with string parameters

### ISSUE THREAD

**Issue by DmitryLobanov** [rexm/Handlebars.Net#141]
Imagine such test added to the `PartialTests` class:

```
        [Test]
        public void BasicPartialWithStringParametersAndImplicitContext()
        {
            string source = "Hello, {{>person last='Smith'}}!";

            var template = Handlebars.Compile(source);

            var data = new
            {
                name = "Marc"
            };

            var partialSource = "{{name}} {{last}}";
            using (var reader = new StringReader(partialSource))
            {
                var partialTemplate = Handlebars.Compile(reader);
                Handlebars.RegisterTemplate("person", partialTemplate);
            }

            var result = template(data);
            Assert.AreEqual("Hello, Marc Smith!", result);
        }
```

This test fails, because partial fails to get parameter `name` from context. But, according to the example on the [Handlebars](http://handlebarsjs.com/) page (section "Partials") this should work. So it seems that if we pass string parameters to partial template, it "forgets" about the context.

Oh, also, if we have partial called from that partial, they both don't get the context. But if we remove string parameters, then context gets passed both to partial and even into partial, included into the first partial.

**Comment by rexm** [comment]
Thanks for the bug report. I don't have a timeframe for a fix yet; but PR always welcome.

**Comment by tsliang** [comment]
**tl;dr Partials don't access their parent BindingContext, and it can be fixed in PathBinder (less effort, possible side effects), PartialBinder (much more effort, but with no side effects), or PartialConverter (medium effort, probably no side effects)**

I've been investigating this issue (as it affects my project as well), and I've come up with the root of the issue, along with possible solutions. 

The problem starts in PartialBinder.VisitPartialExpression:

```csharp
protected override Expression VisitPartialExpression(PartialExpression pex)
        {
            Expression bindingContext = CompilationContext.BindingContext;
            if (pex.Argument != null) // In this test, this is a dict of {"last": "Smith"}
            {
               // Binding context becomes {"last": "Smith"}, with the root BindingContext as a parent
                bindingContext = Expression.Call(
                    bindingContext,
                    typeof(BindingContext).GetMethod("CreateChildContext"),
                    pex.Argument); 
            }
            return Expression.Call(
                new Action<string, BindingContext, HandlebarsConfiguration>(InvokePartial).Method,
                Expression.Convert(pex.PartialName, typeof(string)),
                bindingContext,
                Expression.Constant(CompilationContext.Configuration));
        }
```
We now have a BindingContext that has the capability to fall back to the parent scope. However, that never happens, because when the partial is evaluated, we go into PathBinder.ResolvePath: 

```csharp
        //@param context: the context from the previous example
        //@param path: "name"
        private object ResolvePath(BindingContext context, string path)
        {
            var instance = context.Value; // This is our dictionary of {"last": "Smith"}
            foreach (var segment in path.Split('/'))
            {
                if (segment == "..")
                {
                    context = context.ParentContext; // Only explore parent context when path uses ".."
                    if (context == null)
                    {
                        throw new HandlebarsCompilerException("Path expression tried to reference parent of root");
                    }
                    instance = context.Value;
                }
                else
                {
                    foreach (var memberName in segment.Split('.'))
                    {
                        instance = this.ResolveValue(context, instance, memberName);
                        if (instance is UndefinedBindingResult)
                        {
                            break;
                        }
                    }
                }
            }
            return instance;
        }
```
So what we end up with is a binding context that CAN fall back to the root context, but does not. This leads me to propose three solutions:

1) Change the PathBinder to automatically fall back on a parent BindingContext, no matter what. This is certainly an imperfect solution, because it means that any and all child contexts will fall back to their parents, even if you don't explicitly use a parent path, But I've already tried it out and all tests pass, including @DmitryLobanov's.


2) Find a way to merge BindingContexts so that we can merge the parent context and arguments coming in from the context. I have no idea how you would do this at this level, but it would certainly localize the problem.

3) Find a place in the PartialConverter (maybe AccumulateArguments or so) where we can merge the partial arguments with the root BindingContext. I'm also not sure how doable this is, but it would almost certainly be less work than *#*2, and it would also be (almost certainly) side effect-free

I'm happy to patch whichever option we decide to go with. We're already using *#*1, since the side effects don't affect our templates, but I can understand wanting a more robust solution (and would be happy to do so). Let me know!

**Comment by rexm** [comment]
On the face of it, it seems like the correct solution is when creating the binding for the partial, also create a new sibling context, which is the result of merging the parameters onto a copy of the existing context.

**Comment by tsliang** [comment]
Sure, that makes sense. How do you propose we go about this? I think the easiest way would be to define a new type of class like "MergedParameters" or something like that, then have the PathBinder handle it via a run-time check. Is there a better alternative? I'm still not super-familiar with the codebase ;)

**Comment by rexm** [comment]
Definitely don't want to add special cases to the PathBinder for this. We should be able to create a derived context object in the partial binder and transparently replace...

**Comment by rexm** [comment]
You already found the right place in the code to fix it - [this block](https://github.com/rexm/Handlebars.Net/blob/master/source/Handlebars/Compiler/Translation/Expression/PartialBinder.cs#L34) needs to be a little smarter to create a merged context instead of a child context.

**Comment by tsliang** [comment]
Nah, not always. If you give a partial a child context scope it'll use it; if not, it'll use the parent context. Interestingly, Handlebars will fall back to the parent context if you give a partial a parameter (as in @DmitryLobanov's test case), but it won't if you give it an incomplete child scope, i.e. an object that is missing a value for a partial's template. Examples: http://jsfiddle.net/L6798/28/

**Comment by rexm** [comment]
Good, that looks like it is aligned with how it already works in this library, with the exception of the merged case.

**Comment by amaclean** [comment]
Hey guys, been meaning to fix this for a while as it's been annoying me as well.

Just wanted to offer up a slight alternative to the first solution @tsliang proposed. We could reduce any unintentional side effects of this change by replacing the hash parameter Dictionary with a new class that simply inherits from a Dictionary, then use this class (via a type check on context.value) to decide if we fallback to the parent context in ResolvePath? 

This would be a simple change that keeps functionality inline with Handlebars.

**Comment by rexm** [comment]
@amaclean I've iterated on a few variations of this and none of them are satisfactory... they end up scattering ugly conditionals over a large cross section of the code. I'm not sure my original suggestion above will actually work either. Any suggestions for an elegant and simple solution are welcome...

**Comment by amaclean** [comment]
Just created a PR with a simple fix but I know you weren't keen on making changes to the PathBinder. Let me know if you don't think this is appropriate and we can continue to look for alternatives.

**Comment by tsliang** [comment]
@amaclean's approach is probably the best approach to move forward. I was playing around with merging objects as suggested by @rexm, and it doesn't really lend itself well to the current situation - merging an object must be done at template run-time, which means that the argument to CreateChildContext set on [this line](https://github.com/rexm/Handlebars.Net/blob/master/source/Handlebars/Compiler/Translation/Expression/PartialBinder.cs#L43) is no longer a HashParametersExpression, which leads to path resolution problems. I think it can be done, but if @amaclean's solution works out then it's probably a good idea to run with that until we figure something out with this dilemma.

**Comment by tsliang** [comment]
If you're curious about how I'm approaching this, btw, my work so far is here: https://github.com/tsliang/Handlebars.Net/tree/handlebars-141

I'd appreciate your thoughts!

**Comment by tsliang** [comment]
And of course, a few minutes after I talk about how difficult it is, I found a way to make it work :D See PR #150

**Comment by rexm** [comment]
This was fixed in #149
